### PR TITLE
[SPARK-29688][SQL] Support average for interval type values

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

avg aggregate support interval type values

### Why are the changes needed?

Part of SPARK-27764 Feature Parity between PostgreSQL and Spark

### Does this PR introduce any user-facing change?

yes, we can do avg on intervals

### How was this patch tested?

add ut

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @wangyum @gatorsmile

**Comment by maropu** [comment]
also cc: @MaxGekk

**Comment by maropu** [comment]
@yaooqinn Can you fix the build failure first?

**Comment by yaooqinn** [comment]
@maropu thanks for your review, I have had it rebased with newest master branch, thanks. I appreciate for your time to take another look.

**Comment by maropu** [comment]
btw, this pr includes SPARK-29387, too?

**Comment by yaooqinn** [comment]
Yes, avg needs interval divide feature support.

**Comment by yaooqinn** [comment]
https://github.com/apache/spark/pull/26345 I proposed this to support multiply and divide without noticing spark-29378，@maropu，intended to make these step by step

**Comment by maropu** [comment]
I will check this again after #26132 merged.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 1]
```diff
@@ -97,6 +97,16 @@ object IntervalUtils {
...
   }
 
+  def divide(interval: CalendarInterval, divisor: Decimal): CalendarInterval = {
+    if (divisor == Decimal.ZERO || divisor == null) return null
+    val months = Decimal(interval.months) / divisor
+    val days = months.remainder(Decimal.ONE) * Decimal(DAYS_PER_MONTH) +
+      Decimal(interval.days) / divisor
+    val milliseconds = days.remainder(Decimal.ONE) * Decimal(DateTimeUtils.MICROS_PER_DAY) +
+      Decimal(interval.microseconds) / divisor
+    new CalendarInterval(months.toInt, days.toInt, milliseconds.toLong)
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala]: `milliseconds`?

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala]: microseconds. thanks.

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala]: this won't be needed when your pr will be merged. I'd like to leave it for a while.

**Comment by yaooqinn** [comment]
ping @maropu @MaxGekk @cloud-fan, this have been rebased with lastest interval divide behavior change, thanks very much for review one more time, sorry for the delay.

File: sql/core/src/test/resources/sql-tests/inputs/group-by.sql [hunk 2]
```diff
@@ -192,4 +192,36 @@ having sv is not null;
 SELECT
     i,
     Sum(cast(v as interval)) OVER (ORDER BY i ROWS BETWEEN CURRENT ROW AND UNBOUNDED FOLLOWING)
-FROM VALUES(1,'1 seconds'),(1,'2 seconds'),(2,NULL),(2,NULL) t(i,v);
\ No newline at end of file
+FROM VALUES(1,'1 seconds'),(1,'2 seconds'),(2,NULL),(2,NULL) t(i,v);
+
+-- average with interval type
+-- null
+select avg(cast(v as interval)) from VALUES ('1 seconds'), ('2 seconds'), (null) t(v) where v is null;
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/group-by.sql]: is it the same with `select avg(cast(v as interval)) from VALUES (null) t(v)`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Average.scala [hunk 3]
```diff
@@ -52,11 +51,13 @@ case class Average(child: Expression) extends DeclarativeAggregate with Implicit
   private lazy val resultType = child.dataType match {
     case DecimalType.Fixed(p, s) =>
       DecimalType.bounded(p + 4, s + 4)
+    case CalendarIntervalType => CalendarIntervalType
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Average.scala]: nit: `case interval: CalendarIntervalType => interval`?

**Review by maropu — APPROVED** [review]

**Comment by yaooqinn** [comment]
@cloud-fan @maropu thanks for your review, I will update with the minor revision in a minute. Sorry for the delay, focused on the spark `unnest` implementation just now

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Average.scala (hunk 3 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Average.scala]: ok, I will change this and the one in the sumDataType

File: sql/core/src/test/resources/sql-tests/inputs/group-by.sql (hunk 2 above)
Comment (yaooqinn) [reply sql/core/src/test/resources/sql-tests/inputs/group-by.sql]: yes, I changed it in your way. thanks.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**